### PR TITLE
[SPARK-6354][SQL] Replace the plan which is part of cached query

### PR DESCRIPTION
**Pull request by viirya**
Currently we only replace the plan which equals to cached query. This approach can be extended to replace the plan which is part of cached query.

JIRA: https://issues.apache.org/jira/browse/SPARK-6354

### CODE REVIEW

**Comment by marmbrus** [comment]
While doing a more complete query subsumption calculation would be great, in general this is a very hard problem that is very easy to mess up.  You'll need to clearly define what you mean by `part` and there will need to be a lot more test cases showing that all of the edge cases are correct.  The function itself will also need more comments.  In particular, its not clear to me how you can implement such a function generally without understanding the semantics of the arguments to the operator themselves.  Finally such a significant change to such a central system would need to be hidden behind a feature flag.

Ideally a full design would be proposed on the JIRA before more coding is done.

**Comment by viirya** [comment]
I will update the design on the JIRA and add comments later.

**Comment by viirya** [comment]
@marmbrus I have updated the design on the JIRA.

**Comment by viirya** [comment]
More comments and tests are added now.

**Comment by viirya** [comment]
Looks like an irrelative failure.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala [hunk 1]
```diff
@@ -81,6 +81,39 @@ abstract class LogicalPlan extends QueryPlan[LogicalPlan] with Logging {
...
+   * can do better should override this function.
+   */
+  def partResult(plan: LogicalPlan): Boolean = {
+    plan.getClass == this.getClass &&
+    plan.children.size == children.size && {
+      logDebug(s"[${plan.cleanArgs.mkString(", ")}] is part of [${cleanArgs.mkString(", ")}]")
+      !plan.cleanArgs.zip(cleanArgs).exists {
+        // If this arg is a sequence, we check if there is a sequence in cleanArgs of this logical
+        // plan at the same index that contains all elements of this arg.
+        case (s: Seq[_], ss: Seq[_]) =>
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala]: This is not a valid check to make sure that queries are going to return the same answer.  Consider grouping for example:

`SELECT a FROM table GROUP BY a, b`
`SELECT a FROM table GROUP BY a`

**Comment by marmbrus** [comment]
After thinking about this more, I think this approach is must too simplistic to be correct.  I was able to come up with one trivial example that breaks the logic (and was not caught by tests), but I think this is a really hard problem in general.  Basically equality is easy and can be done with simple plan munging, but doing query subsumption is likely going to require considering each operator carefully.  As a result I think we should close this issue for now until there is a concrete propose for how to handle this correctly in all cases.

**Pull request closed without merging**